src/core/json/path.cc

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "src/core/json/path.h"
+
+#include <absl/types/span.h>
+
+#include "base/expected.hpp"
+#include "base/logging.h"
+#include "src/core/json_object.h"
+
+using namespace std;
+using nonstd::make_unexpected;
+
+namespace dfly::json {
+
+namespace {
+
+template <typename... Ts> struct Overload : Ts... { using Ts::operator()...; };
+
+template <typename... Ts> Overload(Ts...) -> Overload<Ts...>;
+
+bool ShouldIterateAll(SegmentType type) {
+  return type == SegmentType::WILDCARD;
+}
+
+class Dfs {
+ public:
+  using Cb = std::function<void(const JsonType&)>;
+
+  // TODO: for some operations we need to know the type of mismatches.
+  void Traverse(absl::Span<const PathSegment> path, const JsonType& json, const Cb& callback);
+
+  unsigned matches() const {
+    return matches_;
+  }
+
+ private:
+  enum MatchStatus {
+    OUT_OF_BOUNDS,
+    MISMATCH,
+  };
+
+  bool TraverseImpl(absl::Span<const PathSegment> path, const Cb& callback);
+
+  nonstd::expected<void, MatchStatus> PerformStep(const PathSegment& segment, const JsonType& node,
+                                                  const Cb& callback);
+
+  void DoCall(const Cb& callback, const JsonType& node) {
+    ++matches_;
+    callback(node);
+  }
+
+  using AdvanceResult = nonstd::expected<const JsonType*, MatchStatus>;
+
+  class Item {
```

Comment:
A few comments what those classes contain wouldn't hurt 🙂 Because otherwise you have to figure it out based on their usage